### PR TITLE
Add Docker-CE default rules

### PR DESCRIPTION
**Pull request by luisfdez**
#### Add Docker-CE default rules

This Pull Request adds the set of rules and chains that docker-ce daemon sets on a CentOS8 machine by default.

These rules are defined in `ip-filter`, not playing nicely with the defaults shipped for `inet-filter`.

This merge request adds the set of rules that Docker deploys by default. I don't know how relevant all of them are, but I just followed what the daemon was doing.

#### This Pull Request (PR) fixes the following issues

None

### CODE REVIEW

**Comment by nbarrientos** [comment]
Thanks!

Could you please amend the lint errors the CI suggests and also add the new class to the [`all_rules` acceptance test](https://github.com/voxpupuli/puppet-nftables/blob/master/spec/acceptance/all_rules_spec.rb)?

**Comment by luisfdez** [comment]
Thanks Nacho, I have fixed the lint errors and added the new class to the acceptance tests.

**Comment by nbarrientos** [comment]
Thanks! Maybe it'd be a good idea to add some unit test coverage to the new class to help us catching regressions in the future and to make sure for instance that the parameters are correctly taken into account. I've written some that you can just import here if you like them or use as skeleton for your own:

https://github.com/cernops/puppet-nftables/commit/2c53a270ec92efc454c6824aa94eebfe7d6183f0

File: manifests/rules/docker_ce.pp [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  }
+
+  nftables::rule {
+    'POSTROUTING-docker':
+      table   => 'ip-nat',
+      content => "oifname != \"${docker_interface}\" ip saddr ${docker_prefix} counter masquerade";
+    'PREROUTING-docker':
+      table   => 'ip-nat',
+      content => 'fib daddr type local counter jump DOCKER';
+    'OUTPUT-jump_docker':
```
Comment (nbarrientos) [inline manifests/rules/docker_ce.pp:93]:
```suggestion
    'OUTPUT-jump_docker@ip-nat':
```
There's an `OUTPUT` chain in `inet-filter`.

File: manifests/rules/docker_ce.pp [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+#
+# When using this class 'docker::iptables: false' should be set.
+#
+# @param docker_interface
+#   Interface name used by docker. It defaults to docker0.
+# @param docker_prefix
+#   The address space used by docker. It defaults to 172.17.0.0/16.
+#
+class nftables::rules::docker_ce (
+  String                        $docker_interface = 'docker0',
```
Comment (traylenator) [inline manifests/rules/docker_ce.pp:17]:
```suggestion
  String[1]                        $docker_interface = 'docker0',
```
string needs to have length.

**Review by traylenator — CHANGES_REQUESTED** [review]

File: manifests/rules/docker_ce.pp (hunk 2 above)
Comment (luisfdez) [reply manifests/rules/docker_ce.pp]: done.

File: manifests/rules/docker_ce.pp (hunk 1 above)
Comment (luisfdez) [reply manifests/rules/docker_ce.pp]: Thanks, just changed that. It should work both ways, but it does help with readability.

**Comment by luisfdez** [comment]
> Thanks! Maybe it'd be a good idea to add some unit test coverage to the new class to help us catching regressions in the future and to make sure for instance that the parameters are correctly taken into account. I've written some that you can just import here if you like them or use as skeleton for your own:
> 
> [cernops@2b7935f](https://github.com/cernops/puppet-nftables/commit/2b7935f17d9f84447644db88361e7e1a905eb54b)

Thanks Nacho, I've taken your example and expanded with other rules.

**Comment by luisfdez** [comment]
> Presumably as machines are created and destroyed new rules will be created and deleted ?
> Do we need some docs to to set the `noflush_tables` parameter ?

I don't know what would be the right approach to make it rock solid. In principle in a default/basic usage there shouldn't be rules added as it relies on the interface and the address space. In any case, I'm sure that there will be cases not covered.

**Review by traylenator — DISMISSED** [review]

**Comment by tmanninger** [comment]
If the OUTPUT-NAT chain is already defined, i get the a duplication declaration error.
The declaration should be optional.

File: manifests/rules/docker_ce.pp [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+# The configuration distributed in this class represents the default firewall
+# configuration done by docker-ce when the iptables integration is enabled.
+#
+# This class is needed as the default docker-ce rules added to ip-filter conflict
+# with the inet-filter forward rules set by default in this module.
+#
+# When using this class 'docker::iptables: false' should be set.
+#
+# @param docker_interface
+#   Interface name used by docker. It defaults to docker0.
```
Comment (nbarrientos) [inline manifests/rules/docker_ce.pp:12]:
```suggestion
#   Interface name used by docker.
```

The default value will be automatically picked up from the code when generating `REFERENCE.md` so there's no need to specify it here again.

File: manifests/rules/docker_ce.pp [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+# @param docker_prefix
+#   The address space used by docker. It defaults to 172.17.0.0/16.
```
Comment (nbarrientos) [inline manifests/rules/docker_ce.pp:14]:
```suggestion
#   The address space used by docker.
```
The default value will be automatically picked up from the code when generating `REFERENCE.md` so there's no need to specify it here again.

**Review by nbarrientos — CHANGES_REQUESTED** [review]

File: manifests/rules/docker_ce.pp (hunk 3 above)
Comment (luisfdez) [reply manifests/rules/docker_ce.pp]: Done.

File: manifests/rules/docker_ce.pp (hunk 4 above)
Comment (luisfdez) [reply manifests/rules/docker_ce.pp]: Done.

File: manifests/rules/docker_ce.pp [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+      'DOCKER_ISOLATION_STAGE_1': ;
+      'DOCKER_ISOLATION_STAGE_2': ;
+      'DOCKER_USER': ;
+    }
+  }
+
+  nftables::rule {
+    'DOCKER_ISOLATION_STAGE_1-iifname':
+      order   => '01',
+      content => "iifname \"${docker_interface}\" oifname != \"${docker_interface}\" counter jump DOCKER_ISOLATION_STAGE_2";
```
Comment (bastelfreak) [inline manifests/rules/docker_ce.pp:40]: same question as in the other PR: are the escaped quotes required / can we use single quotes?

**Comment by luisfdez** [comment]
> If the OUTPUT-NAT chain is already defined, i get the a duplication declaration error.
> The declaration should be optional.

@tmanninger I have added a couple of class parameters to handle the creation of chains. Does it look good to you now? Thanks for taking a look :-)

File: manifests/rules/docker_ce.pp (hunk 5 above)
Comment (luisfdez) [reply manifests/rules/docker_ce.pp:40]: Normal interface names: eth0, docker0 could go without double quotes. However, if for some reason some uncommon interface name is used double quotes are needed. This approach should be safer for those uncommon cases.
There is an example of this in `man nft` describing the string type.

**Review by bastelfreak — APPROVED** [review]

**Review by nbarrientos — APPROVED** [review]

**Review by duritong — APPROVED** [review]

**Pull request merged**